scripts/tests/sanitycheck/test_testsuite_class.py

```diff
@@ -38,3 +41,170 @@ def test_testsuite_add_testcases(class_testsuite):
+        failed_platform_list = [row["platform"] for row in csv.DictReader(filepath) if row["status"] == "failed"]
+        for row in failed_platform_list:
+            with patch.object(TestSuite, 'get_platform') as mock_method:
+                class_testsuite.load_from_file(class_testsuite.outdir + "sanitycheck.csv", filter_status=["Skipped", "Passed"])
+                calls = [call(row)]
+                mock_method.assert_has_calls(calls, any_order=True)
+                assert mock_method.call_count == len(failed_platform_list)
+
+    # Scenario 4 : Assert add_instances function is called from load_from_file function
+    class_testsuite.add_instances = MagicMock(side_effect=class_testsuite.add_instances)
+    class_testsuite.load_from_file(test_data + "sanitycheck.csv")
+    class_testsuite.add_instances.assert_called()
+
+    # Scenario 5 : Validate if the Keyerror is raised in case if a header expected is missing
+    with pytest.raises(SystemExit):
+        class_testsuite.load_from_file(test_data + "sanitycheck_keyerror.csv")
+    assert "Key error while parsing tests file.('status')" in caplog.text
+
+TESTDATA_PART1 = [
+    ("toolchain_whitelist", ['gcc'], None, None, "Not in testcase toolchain whitelist"),
+    ("platform_whitelist", ['demo_board_1'], None, None, "Not in testcase platform whitelist"),
+    ("toolchain_exclude", ['zephyr'], None, None, "In test case toolchain exclude"),
+    ("platform_exclude", ['demo_board_2'], None, None, "In test case platform exclude"),
+    ("arch_exclude", ['x86_demo'], None, None, "In test case arch exclude"),
+    ("arch_whitelist", ['arm'], None, None, "Not in test case arch whitelist"),
+    ("skip", True, None, None, "Skip filter"),
+    ("tags", set(['sensor', 'bluetooth']), "ignore_tags", ['bluetooth'], "Excluded tags per platform"),
+    ("min_flash", "2024", "flash", "1024", "Not enough FLASH"),
+    ("min_ram", "500", "ram", "256", "Not enough RAM"),
+    ("None", "None", "env", ['BSIM_OUT_PATH', 'demo_env'], "Environment (BSIM_OUT_PATH, demo_env) not satisfied"),
+    ("build_on_all", True, None, None, "Platform is excluded on command line."),
+    (None, None, "supported_toolchains", ['gcc'], "Not supported by the toolchain")
+]
+
+
+@pytest.mark.parametrize("tc_attribute, tc_value, plat_attribute, plat_value, expected_discards", TESTDATA_PART1)
+def test_apply_filters_part1(class_testsuite, all_testcases_dict, platforms_list,
+                             tc_attribute, tc_value, plat_attribute, plat_value, expected_discards):
+    """ Testing apply_filters function of TestSuite class in Sanitycheck
+    Part 1: Response of apply_filters function (discard dictionary) have
+    appropriate values according to the filters
+    """
+    class_testsuite.platforms = platforms_list
+    class_testsuite.testcases = all_testcases_dict
+    for plat in class_testsuite.platforms:
```

Comment:
does not look great and does not have as much coverage as we would want to have, for example for ram restrictions, but that is fine for now...

Reply:
Does the test data line #147 in TESTDATA_PART1 not cover the RAM restriction? Am i missing something here? 

'''
("min_ram", "500", "ram", "256", "Not enough RAM"),
'''

Reply:
look at https://github.com/zephyrproject-rtos/zephyr/pull/24518/commits/f88678c13db337fe5c44407af7867212eaa02d39#diff-6fbb35ecd03a71dc2d629502d7d18607R33, the last test in the file as an example...

Reply:
Added another testcase "test_apply_filters_part3: to test the edge cases for ram and flash values.  Let me know if it looks fine. Thanks